RecoLocalCalo/HcalRecAlgos/src/DoMahiAlgo.cc

```diff
@@ -0,0 +1,563 @@
+#include "RecoLocalCalo/HcalRecAlgos/interface/DoMahiAlgo.h" 
+#include "FWCore/MessageLogger/interface/MessageLogger.h"
+#include <iostream>
+#include <fstream> 
+
+void eigen_solve_submatrix(PulseMatrix& mat, PulseVector& invec, PulseVector& outvec, unsigned NP);
+
+DoMahiAlgo::DoMahiAlgo() :
+  FullTSSize_(19), //19
+  FullTSofInterest_(8) //8
+{}
+
+void DoMahiAlgo::setParameters(double iTS4Thresh, double chiSqSwitch, bool iApplyTimeSlew, HcalTimeSlew::BiasSetting slewFlavor,
+			       double iMeanTime, double iTimeSigmaHPD, double iTimeSigmaSiPM, 
+			       const std::vector <int> &iActiveBXs, int iNMaxItersMin, int iNMaxItersNNLS,
+			       double iDeltaChiSqThresh, double iNnlsThresh) {
+
+  TS4Thresh_     = iTS4Thresh;
+  chiSqSwitch_   = chiSqSwitch;
+
+  applyTimeSlew_ = iApplyTimeSlew;
+  slewFlavor_    = slewFlavor;
+
+  meanTime_      = iMeanTime;
+  timeSigmaHPD_  = iTimeSigmaHPD;
+  timeSigmaSiPM_ = iTimeSigmaSiPM;
+
+  activeBXs_ = iActiveBXs;
+
+  nMaxItersMin_  = iNMaxItersMin;
+  nMaxItersNNLS_ = iNMaxItersNNLS;
+
+  deltaChiSqThresh_ = iDeltaChiSqThresh;
+  nnlsThresh_    = iNnlsThresh;
+
+  BXOffsetConf_ = -(*std::min_element(activeBXs_.begin(), activeBXs_.end()));
+  BXSizeConf_   = activeBXs_.size();
+}
+
+void DoMahiAlgo::phase1Apply(const HBHEChannelInfo& channelData,
+			     float& reconstructedEnergy,
+			     float& reconstructedTime,
+			     float& chi2) {
+
+  TSSize_ = channelData.nSamples();
+  TSOffset_ = channelData.soi();
+  FullTSOffset_ = FullTSofInterest_ - TSOffset_;
+
+  // 1 sigma time constraint
+  if (channelData.hasTimeInfo()) dt_=timeSigmaSiPM_;
+  else dt_=timeSigmaHPD_;
+
+  //fC to photo-electron scale factor (for P.E. uncertainties)
+  fcByPe_ = channelData.fcByPE();
+
+  //Dark current value for this channel (SiPM only)
+  darkCurrent_ =  getSiPMDarkCurrent(channelData.darkCurrent(), 
+				     channelData.fcByPE(),
+				     channelData.lambda());
+
+  //Average pedestal width (for covariance matrix constraint)
+  pedConstraint_ = 0.25*( channelData.tsPedestalWidth(0)*channelData.tsPedestalWidth(0)+
+			  channelData.tsPedestalWidth(1)*channelData.tsPedestalWidth(1)+
+			  channelData.tsPedestalWidth(2)*channelData.tsPedestalWidth(2)+
+			  channelData.tsPedestalWidth(3)*channelData.tsPedestalWidth(3) );
+
+  std::vector<float> reconstructedVals;
+  SampleVector charges;
+  
+  double tsTOT = 0, tstrig = 0; // in GeV
+  for(unsigned int iTS=0; iTS<TSSize_; ++iTS){
+    double charge = channelData.tsRawCharge(iTS);
+    double ped = channelData.tsPedestal(iTS);
+
+    charges.coeffRef(iTS) = charge - ped;
+
+    //ADC granularity
+    double noiseADC = (1./sqrt(12))*channelData.tsDFcPerADC(iTS);
```

Comment:
it may help to encapsulate this in HBHEChannelInfo and define a noiseADC method there (to replace the use case in PulseShapeFitOOTPileupCorrection as well).

Reply:
We would prefer to defer this change to a later PR, and in fact add the whole sum computed here (https://github.com/jaylawhorn/cmssw/blob/016fc04ee6d5908665a3ef42652b926c624ffce2/RecoLocalCalo/HcalRecAlgos/src/DoMahiAlgo.cc#L96) as a field to the HBHEChannel info so that it can be accessed by both Mahi and Method 2 in the same way. (@mariadalfonso will do this)

Reply:
OK